### PR TITLE
SCP-2238: Replace eventful with beam

### PR DESCRIPTION
**Pull request by silky**
WIP

TODO:

- [x] (Rebase; it's quite behind)
- [x] Test that it works; build a few scenarios
- [ ] ~Maybe change the DB structure to be a bit less stateful and track history more?~ (n/a?)
- [x] ~Check if it's faster than eventful~ ( it's not _faster_, but it uses less memory and db is smaller; further investigating to be done in a different issue)
- [ ] ~Add a benchmark for this?~ ( todo in a later issue )
- [ ] ~Remove specialisation to Sqlite in `DbStore` effect?~
- [x] Clean up todo notes/`error`s
- [x] ~Make the `migrate` command use the settings from the config file ( revert this - https://github.com/input-output-hk/plutus/pull/3158/files ?? )~
- [x] ~Delete all eventful code~

----

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-pab/src/Plutus/PAB/Db/Beam/ContractDefinitionStore.hs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+mkRow (ContractExe {..}) = Contract (Text.pack contractPath)
+
+fromRow :: Contract -> ContractExe
+fromRow (Contract {..})  = ContractExe (Text.unpack _contractPath)
+
+handleContractDefinitionStore ::
+  forall effs.
+  ( Member DbStoreEffect effs
+  )
+  => ContractDefinitionStore ContractExe
```
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Db/Beam/ContractDefinitionStore.hs:38]: Ideally we'd have a type parameter instead of `ContractExe` (with some constraints to make the serialisation work), but this probably requires some changes in other places, so maybe it's better to leave it for another PR.

**Review by j-mueller — COMMENTED** [review]
Looking good, this is what I thought it would be like

File: plutus-pab/src/Plutus/PAB/Db/Beam/ContractStore.hs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+          (\ci -> ci ^. contractInstanceActive <-. val_ False)
+          (\ci -> ci ^. contractInstanceId ==. val_ (uuidStr instanceId))
+
+  GetActiveContracts ->
+    fmap mkContracts
+      $ selectList
+      $ select
+      $ do
+          ci <- all_ (_contractInstances db)
+          guard_ ( ci ^. contractInstanceActive ==. val_ True )
```
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Db/Beam/ContractStore.hs:132]: does this work?

```suggestion
          guard_ (ci ^. contractInstanceActive)
```

File: plutus-pab/src/Plutus/PAB/Effects/DbStore.hs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE RankNTypes           #-}
+{-# LANGUAGE TemplateHaskell      #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# options_ghc -Wno-missing-signatures #-}
+
+module Plutus.PAB.Effects.DbStore where
```
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Effects/DbStore.hs:18]: missing module documentation

File: plutus-pab/app/CommandParser.hs [hunk 4]
```diff
@@ -141,6 +130,7 @@ migrationParser =
                 str
                 (metavar "DATABASE" <>
                  help "The sqlite database file.")
+        -- TODO: This will need to be 'WithConfig'.
```
Comment (j-mueller) [inline plutus-pab/app/CommandParser.hs:133]: Please check with @gilligan - being able to run `migrate` without a config file was added to make the VM test easier

File: plutus-pab/app/CommandParser.hs [hunk 5]
```diff
@@ -50,21 +46,14 @@ ekgFlag =
...
         AppOpts <$> logLevelFlag
-               <*> logConfigFileParser
-               <*> configFileParser
-               <*> ekgFlag
-               <*> inMemoryFlag
-               <*> commandParser
+                <*> logConfigFileParser
+                <*> configFileParser
+                <*> ekgFlag
+                -- <*> dbKindParser
```
Comment (j-mueller) [inline plutus-pab/app/CommandParser.hs:55]: delete?

File: plutus-pab/put-data-in-pab.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
+#!/bin/env bash
```
Comment (j-mueller) [inline plutus-pab/put-data-in-pab.sh:1]: not sure what to do about the two scripts - it would be nice to have them in the repo, but they might get out of date really quickly if we just put them like this. The best approach IMO would be to add them as nix scripts and then run them as part of the VM tests. This could be a separate PR though.

**Review by j-mueller — APPROVED** [review]

File: plutus-pab/src/Plutus/PAB/Db/Beam/ContractStore.hs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  , Member (Error PABError) effs
+  )
+  => ContractStore ContractExe
+  ~> Eff effs
+handleContractStore = \case
+  PutStartInstance args instanceId ->
+    addRow (_contractInstances db)
+      $ mkRow args instanceId
+
+  -- TODO: Should we use 'args' ?
```
Comment (j-mueller) [inline plutus-pab/src/Plutus/PAB/Db/Beam/ContractStore.hs:95]: It doesn't change over the lifetime so we don't need to update it

File: plutus-pab/put-data-in-pab.sh (hunk 6 above)
Comment (silky) [reply plutus-pab/put-data-in-pab.sh:1]: sounds good @j-mueller :)

i'd also be happy to temporarily put them in a `hacky-scripts` folder to imply that they aren't expected to be long-lived/supported.

File: plutus-pab/app/CommandParser.hs (hunk 4 above)
Comment (silky) [reply plutus-pab/app/CommandParser.hs:133]: @j-mueller I see. At the time of this comment I had anticpated needing more information in the config in order to even interact with the database; but now I don't require it, so I'm happy to leave it as-is if it's working.

**Comment by silky** [comment]
hey @j-mueller ; there's one warning hanging around that i left in to remind myself of one fact:

i noticed in the original code we had a handling for in-memory (_not_ via eventful) storage of the contracts and instances; but, i noticed that it _only_ happened for the instances (the contract definitions were _always_ stored in eventful)

i silently dropped anything to do with the in-memory part ( https://github.com/input-output-hk/plutus/pull/3294/files#diff-591cd2da7d1bea2816b5c929fb9b77e9206f30e5a4ccf33c0f333234903414acL101 )

i think i should add this back in, _and_ add in a step to store the contract _definitions_ in memory as well? or should i just put it back as it was?

**Comment by j-mueller** [comment]
We can't store the contract definitions in-memory because they are installed using separate CLI invocations (in separate processes) and the memory isn't preserved between them.

So I'd just put it back as it was.

Also note that when we have fully switched to the `Builtin` way of running contracts, as opposed to `ContractExe`, then we won't need to store the definitions at all because they are defined as a Haskell datatype. We could drop that table completely then.

**Comment by silky** [comment]
Ah okay, fair enough, that makes total sense, thanks @j-mueller ! :)

File: plutus-pab/put-data-in-pab.sh (hunk 6 above)
Comment (j-mueller) [reply plutus-pab/put-data-in-pab.sh:1]: Ok, since you're going to do the VM tests right after this one we can leave the scripts in for now.

**Review by j-mueller — APPROVED** [review]

File: plutus-pab/put-data-in-pab.sh (hunk 6 above)
Comment (silky) [reply plutus-pab/put-data-in-pab.sh:1]: https://jira.iohk.io/browse/SCP-2385
